### PR TITLE
fix: use local time for time buckets and improve memories

### PR DESCRIPTION
**Pull request by jrasm91**
- save local date and time in a new column (`exifInfo.localDateTime`). This new column stores the local capture date and time as the UTC timezone
- update time bucket queries to use local date time
- update group date rendering to group by local date time
- update group date titles to use local date time
- memory lane finds memories from all years, not just 30 years ago and is faster
- memory lane uses local date to determine correct date to display. This way, photos taken in the evening in a western timezone shows the correct date even if viewed in an eastern timezone

**TODO**
- memory lane viewer on mobile does not use local date time yet, can fix later
-  ~~update video extraction to save local date time~~, @jrasm91 is this still needed?

**Screenshots**
Here, we have a bunch of photos taken in the US on the evening of April 5, plus a photo taken in Europe on the morning of the 6th and a photo taken in the US on morning on the 6th. We are looking at the web page from a European time zone.

Before PR, they are all put in the April 6 bucket even though two photos are taken on the 5th:
image

After PR, they are all put in the correct day according to local time:
image

Note that strictly speaking the photo of the red car is taken _before_ the evening photos on April 5, but in a different timezone. 

Memory view from image taken 58 years ago. This photo was taken in the US on the evening of the 19th (exif data obviously faked). 
image

### CODE REVIEW

**Review by etnoy — APPROVED** [review]

**Comment by alextran1502** [comment]
Can we fix the merge conflict? I can help with testing

**Comment by etnoy** [comment]
> Can we fix the merge conflict? I can help with testing

done!

File: server/src/domain/metadata/metadata.service.ts [hunk 1]
```diff
@@ -156,9 +157,18 @@ export class MetadataService {
     await this.applyMotionPhotos(asset, tags);
     await this.applyReverseGeocoding(asset, exifData);
     await this.assetRepository.upsertExif(exifData);
+    let localDateTime = exifData.dateTimeOriginal ?? undefined;
+
+    const dateTimeOriginal = exifDate(firstDateTime(tags as Tags)) ?? exifData.dateTimeOriginal;
+    const timeZoneOffset = tzOffset(firstDateTime(tags as Tags)) ?? 0;
+
+    if (dateTimeOriginal && timeZoneOffset) {
+      localDateTime = new Date(dateTimeOriginal.getTime() + timeZoneOffset * 60000);
```
Comment (jrasm91) [inline server/src/domain/metadata/metadata.service.ts:166]: This looks good. Doing it with luxon probably would be cleaner, but it's probably good enough.

File: server/src/infra/migrations/1694525143117-AddLocalDateTime.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`ALTER TABLE "assets" ADD "localDateTime" TIMESTAMP`);
+    await queryRunner.query(`
+      update "assets"
+        set "localDateTime" = "fileCreatedAt"`);
+
+    await queryRunner.query(`ALTER TABLE "assets" ALTER COLUMN "localDateTime" SET NOT NULL`);
+    await queryRunner.query(`CREATE INDEX "IDX_day_of_month" ON assets (EXTRACT(DAY FROM "localDateTime"))`);
+    await queryRunner.query(`CREATE INDEX "IDX_month" ON assets (EXTRACT(MONTH FROM "localDateTime"))`);
+    await queryRunner.query(`CREATE INDEX "IDX_year" ON assets (EXTRACT(YEAR FROM "localDateTime"))`);
```
Comment (jrasm91) [inline server/src/infra/migrations/1694525143117-AddLocalDateTime.ts:15]: Looks you forgot to remove this one.

File: server/src/infra/repositories/asset.repository.ts [hunk 3]
```diff
@@ -454,8 +481,9 @@ export class AssetRepository implements IAssetRepository {
   getByTimeBucket(timeBucket: string, options: TimeBucketOptions): Promise<AssetEntity[]> {
     const truncateValue = truncateMap[options.size];
     return this.getBuilder(options)
-      .andWhere(`date_trunc('${truncateValue}', "fileCreatedAt") = :timeBucket`, { timeBucket })
-      .orderBy('asset.fileCreatedAt', 'DESC')
+      .andWhere(`date_trunc('${truncateValue}', "localDateTime") = :timeBucket`, { timeBucket })
+      .orderBy(`date_trunc('day', "localDateTime")`, 'DESC')
+      .addOrderBy('asset.localDateTime', 'DESC')
```
Comment (jrasm91) [inline server/src/infra/repositories/asset.repository.ts:486]: Why do we need both of these? It seems like one would be sufficient. Maybe would would need both if you were sorting by `fileCreatedAt` somewhere.

File: server/src/infra/migrations/1694525143117-AddLocalDateTime.ts (hunk 2 above)
Comment (etnoy) [reply server/src/infra/migrations/1694525143117-AddLocalDateTime.ts:15]: the year? yep, forgot it, now removed

File: server/src/infra/repositories/asset.repository.ts (hunk 3 above)
Comment (etnoy) [reply server/src/infra/repositories/asset.repository.ts:486]: Ah, I am sorting the second one on the wrong field. Will correct to fileCreatedAt. I think.

File: server/src/domain/metadata/metadata.service.ts (hunk 1 above)
Comment (etnoy) [reply server/src/domain/metadata/metadata.service.ts:166]: Maybe I'll use luxon later on

**Comment by jrasm91** [comment]
Tested and everything seems to be in order! Thanks for the help @etnoy.

**Comment by waclaw66** [comment]
This PR causes showing no photos at all, just empty Photos page. Albums are visible, but no photos inside.
Photos page request `/api/asset/time-bucket?size=MONTH&isArchived=false&timeBucket=2023-09-30T22:00:00.000Z` returns empty json. No errors within log.

**Comment by etnoy** [comment]
> This PR causes showing no photos at all, just empty Photos page. Albums are visible, but no photos inside. Photos page request `/api/asset/time-bucket?size=MONTH&isArchived=false&timeBucket=2023-09-30T22:00:00.000Z` returns empty json. No errors within log.

Try rerunning all metadata

**Comment by etnoy** [comment]
> This PR causes showing no photos at all, just empty Photos page. Albums are visible, but no photos inside. Photos page request `/api/asset/time-bucket?size=MONTH&isArchived=false&timeBucket=2023-09-30T22:00:00.000Z` returns empty json. No errors within log.

Which timezone are you in?

**Comment by waclaw66** [comment]
> Which timezone are you in?

GMT+2
Europe/Prague

**Comment by waclaw66** [comment]
> 
> Try rerunning all metadata

Trying, but it's extremely slow with 16000+ assets, about 30 assets (8 active) per minute on a i3-12100T.

**Pull request merged**